taskcat the package manager

jaymccon opened the pull request:
## Overview

Initial implementation of taskcat as a package manager. Intention is that these features will be alpha in v0.9.

Implements commands:

* `deploy`
* `list`
* `delete`
* `test list`
* `test clean`

The `deploy` command takes a local path, github org/repo or a quickstart name. for a quickstart or github it checks out the latest master branch of the repo (including submodules) into a local package cache. For example:

```
taskcat deploy ./my-project/
taskcat deploy myorg/myrepo
taskcat deploy redhat-openshift   # equivalent to install aws-quickstart/quickstart-redhat-openshift
```

## Testing/Steps taken to ensure quality

manual execution of the commands against several quick start repos.

### Notes

functionality is in an alpha state, won't focus too much on tests, or impeccable code until there's more certainty around the details of the implementation.

## Testing Instructions

```
taskcat deploy redhat-openshift --region us-east-1
taskcat list
taskcat delete <NAME_OR_UUID> --region us-east-1
```

jaymccon commented:
/do-e2e-tests

andrew-glenn commented on taskcat/_cfn/threaded.py:
```diff
@@ -228,3 +230,55 @@ def _group_stacks(stacks: Stacks) -> List[dict]:
+                        k, v = (tag["Key"], tag["Value"])
+                        if k.startswith("taskcat-"):
+                            stack[k] = v
+                    if stack.get("taskcat-id"):
```
I'm having trouble understanding why you would convert it into a UUID.

jaymccon replied:
just because everything else wants the id as a uuid, but probably not needed here

andrew-glenn commented on taskcat/_cli_modules/delete.py:
```diff
@@ -0,0 +1,53 @@
+
+
+class Delete:
+    """[ALPHA] Deletes an installed package in an AWS account/region"""
```
I have concerns around the term "package".

jaymccon replied:
agree, struggling to come up with decent naming conventions, let's discuss

andrew-glenn replied:
We've spoken and resolved this contention point

andrew-glenn commented on taskcat/_cli_modules/install.py:
```diff
@@ -0,0 +1,171 @@
+            package_type = "local"
+        elif "/" in package:
+            package_type = "github"
+        else:  # assuming it's an AWS Quick Start
```
I also thought we were getting away from aws-quickstart assumptions.

jaymccon replied:
this doesn't enforce anything quickstart specific, it just uses quickstart org as the default "repository" for installation. So install that can be done simply from any github repo:

`taskcat install myorg/myrepo`

andrew-glenn commented on taskcat/_cli_modules/install.py:
```diff
@@ -0,0 +1,171 @@
+            f" {stacks.uid.hex}",
+            extra={"nametag": "\x1b[0;30;47m[INSTALL_ID  ]\x1b[0m"},
+        )
+        LOG.error(f" {name}", extra={"nametag": "\x1b[0;30;47m[INSTALL_NAME]\x1b[0m"})
```
Doing the escape characters (for custom colors) here is not scalable IMO;

jaymccon replied:
yup, total shortcut

andrew-glenn commented on taskcat/_cli_modules/list.py:
```diff
@@ -0,0 +1,93 @@
+                string += " "
+            return string
+
+        header = f"NAME{spaces(longest_name)}ID{spaces(34)}NUMBER_OF_STACKS  REGIOM"
```
I think this breaks the UI changes that @avattathil was working on

jaymccon replied:
will update it once that work has finalized

jaymccon commented:
/do-e2e-tests

jaymccon commented:
@andrew-glenn @avattathil this is ready for another review. I've changed the verb to `deploy` and updated the list output to include the project name, so it now looks like:

```
NAME                 PROJECT                           ID                                  REGION
lumpy-rattlesnake    quickstart-redhat-openshift       67c812465c28449490744c945513c094    us-west-2
```

jaymccon commented:
/do-e2e-tests

jaymccon commented:
/do-e2e-tests

jaymccon commented:
ok, sorry for the noise here, had some issues with my pre-commit, then decided to bundle a bit more changes in here, in the interest of getting some important fixes/changes in soonest.

@andrew-glenn @avattathil REALLY ready for review now 😸 let's get this merged!
